pgjdbc/src/main/java/org/postgresql/util/PasswordUtil.java

```diff
@@ -0,0 +1,220 @@
+  }
+
+  /**
+   * Encode the given password for SCRAM-SHA-256 authentication using the default
+   * iteration count and a random salt.
+   *
+   * @param password The plain text of the user's password
+   * @return The text representation of the password encrypted for SCRAM-SHA-256
+   *         authentication
+   */
+  public static String encodeScramSha256(char[] password) {
+    Objects.requireNonNull(password, "password");
+    try {
+      SecureRandom rng = getSecureRandom();
+      byte[] salt = rng.generateSeed(DEFAULT_SALT_LENGTH);
+      return encodeScramSha256(password, DEFAULT_ITERATIONS, salt);
+    } finally {
+      Arrays.fill(password, ' ');
+    }
+  }
+
+  /**
+   * Encode the the given password for use with md5 authentication. The PostgreSQL
+   * server uses the username as the per-user salt so that must also be provided.
+   * The return value of this method is the literal text that may be used when
+   * creating or modifying a user with the given password without the surrounding
+   * single quotes.
+   *
+   * @param user     The username of the database user
+   * @param password The plain text of the user's password
+   * @return The text representation of the password encrypted for md5
+   *         authentication.
+   */
+  public static String encodeMd5(String user, char[] password) {
+    Objects.requireNonNull(password, "password");
+    ByteBuffer passwordBytes = null;
+    try {
+      passwordBytes = StandardCharsets.UTF_8.encode(CharBuffer.wrap(password));
+      byte[] userBytes = user.getBytes(StandardCharsets.UTF_8);
+      final MessageDigest md = MessageDigest.getInstance("MD5");
+
+      md.update(passwordBytes);
+      md.update(userBytes);
+      byte[] digest = md.digest(); // 16-byte MD5
+
+      final byte[] encodedPassword = new byte[35]; // 3 + 2 x 16
+      encodedPassword[0] = (byte) 'm';
+      encodedPassword[1] = (byte) 'd';
+      encodedPassword[2] = (byte) '5';
+      MD5Digest.bytesToHex(digest, encodedPassword, 3);
+
+      return new String(encodedPassword, StandardCharsets.UTF_8);
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException("Unable to encode password with MD5", e);
+    } finally {
+      Arrays.fill(password, ' ');
+      if (passwordBytes != null) {
+        if (passwordBytes.hasArray()) {
+          Arrays.fill(passwordBytes.array(), (byte) 0);
+        } else {
+          int limit = passwordBytes.limit();
+          for (int i = 0; i < limit; i++) {
+            passwordBytes.put(i, (byte) 0);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Encode the given password using the driver's default encryption method.
+   *
+   * @param user     The username of the database user
+   * @param password The plain text of the user's password
+   * @return The encoded password
+   * @throws SQLException If an error occurs encoding the password
+   */
+  public static String encodePassword(String user, char[] password) throws SQLException {
```

Comment:
It looks like the method is not reliable from the users' perspective.
What does `driver's default` mean?
What if somebody uses the method, and then they upgrade the driver some time later. Is the driver allowed to change the default encoding method?

Apparently, for backward compatibility, we can't change the method. In that regard, `encodePassword` duplicates `encodeScramSha256`.

Reply:
> It looks like the method is not reliable from the users' perspective. What does driver's default mean?

It means the driver encodes the password with whatever the latest version of password encoding the driver wants, using the defaults built into the driver.

> What if somebody uses the method, and then they upgrade the driver some time later. Is the driver allowed to change the default encoding method?

Yes. That's exactly the point. So that code targeting that method uses the latest, most recommended method of encoding passwords without being connected to a specific server.

> Apparently, for backward compatibility, we can't change the method. In that regard, encodePassword duplicates encodeScramSha256.

Compatibility with what? We haven't haven't released anything yet. Are you suggesting changing the signature?

It doesn't duplicate the SCRAM-SHA-256 function, it it delegates to it because that's the current driver default.

If in the future if the SCRAM-SHA-256 default is replaced with SCRAM-SHA-512 or something else entirely, we'd change that delegation.

Reply:
> If in the future if the SCRAM-SHA-256 default is replaced with SCRAM-SHA-512 or something else entirely, we'd change that delegation

If we ever make such a change, then we effectively break backward compatibility. That means we can't easily make such a change.

So, please suggest what is the use case for having "driver's default" `encodePassword` method. Why add the method assuming there's not a single use case for it?

Reply:
> If we ever make such a change, then we effectively break backward compatibility. That means we can't easily make such a change.

We don't break anything because the definition of that method is encoding the password with whatever the driver considers to be the most secure and recommended approach. The user is delegating to this driver, as the de facto Java driver for PosgreSQL, to make a determination of how the user should be encoding passwords.

If a user wants to use a specific algo or parameters then there's other overloads to use instead.

> So, please suggest what is the use case for having "driver's default" encodePassword method. Why add the method assuming there's not a single use case for it?

It's in the original PR description:

>> Splitting out the encoding allows the same functions to be used for CREATE USER ... (again without passing the credentials in plaintext).

A user can leverage that to generate their own SQL that involves encoding passwords.

Reply:
> > If we ever make such a change, then we effectively break backward compatibility. That means we can't easily make such a change.
> 
> We don't break anything because the definition of that method is encoding the password with whatever the driver considers to be the most secure and recommended approach. The user is delegating to this driver, as the de facto Java driver for PosgreSQL, to make a determination of how the user should be encoding passwords.

I think the issue is that if the server uses SCRAM-512 for the latest version and previous versions use SCRAM-256 using the latest driver would fail on older versions of the server.

Reply:
For what is worth, Java’s `MessageDigest` does not have “get default hash algorithm”, and they require users to explicitly specify the algorithm.

Reply:
I think you mean if the _driver_ is bumped to SCRAM-512 and the server does not yet support it, then it would fail.

Yes, that's expected because the output of that method is not for a particular server. It's for generating literals for the encoded password using the latest recommended method per the driver. I'd see it being used by something that is generating it's own SQL, potentially for future execution out of band. The tie in to the driver is that the driver, as the de factor Java driver for PostgreSQL, is aware of the recommended password encoding.

Reply:
I don't follow ... isn't that the same situation I described in my previous comment?

Reply:
I suppose it is, but we can't break backward compatibility. If a user upgrades the driver their code should continue to work. 
Generally if we do a major version upgrade we would mention that we have breaking changes but a breaking change to a `default` seems wrong somehow

Reply:
I guess I don't see it as a break as the functional meaning of the method ("*encode my password using the recommended...*) has not changed. I see it like the following changing over server versions:

```
CREATE USER foo WITH PASSWORD 'abcd1234';
SELECT passwd FROM pg_shadow WHERE usename = 'foo'
```

It'd be md5 encoded in <=10 and some variant SCRAM after that. Though that's not as user facing so maybe not the best example.

The original goal was to have a method that users could rely on when making things like SQL script generating tooling that they know will always be the latest recommendation. That way when SCRAM-SHA-256 is replaced with SCRAM-SHA-512 or something else entirely, a user that bumps their driver to the latest pgjdbc would automatically get the newer recommendation.

If that doesn't make sense as valid use case or if you foresee misuse of it causing complication, then let's remove it.